src/perf_tool_backend.ml

```diff
@@ -113,58 +113,90 @@ module Perf_line = struct
-                | "return" -> Return
-                | "tr strt" -> Start
-                | "tr end" -> End None
-                | "tr end  call" -> End Call
-                | "tr end  return" -> End Return
-                | "tr end  syscall" -> End Syscall
-                | "jmp" -> Jump
-                | "jcc" -> Jump
-                | kind -> raise_s [%message "unrecognized perf event" (kind : string)])
-            ; addr
-            ; symbol
-            ; offset
-            ; ip
-            ; ip_symbol
-            ; ip_offset
+      match Re.Group.all (Re.exec re line) with
+      | [| _
+         ; pid
+         ; tid
+         ; time_hi
+         ; time_lo
+         ; kind
+         ; src_instruction_pointer
+         ; src_symbol_and_offset
+         ; dst_instruction_pointer
+         ; dst_symbol_and_offset
+        |] ->
+        let pid = Int.of_string pid in
+        let tid = Int.of_string tid in
+        let time_lo =
+          (* In practice, [time_lo] seems to always be 9 decimal places, but it seems good to guard
+             against other possibilities. *)
+          let num_decimal_places = String.length time_lo in
+          match Ordering.of_int (Int.compare num_decimal_places 9) with
+          | Less -> Int.of_string time_lo * Int.pow 10 (9 - num_decimal_places)
+          | Equal -> Int.of_string time_lo
+          | Greater -> Int.of_string (String.prefix time_lo 9)
+        in
+        let time_hi = Int.of_string time_hi in
+        let int64_of_hex_string str =
+          try Scanf.sscanf str "%Lx" Fn.id with
+          | Scanf.Scan_failure _ | End_of_file -> 0L
+        in
+        let src_instruction_pointer = int64_of_hex_string src_instruction_pointer in
+        let dst_instruction_pointer = int64_of_hex_string dst_instruction_pointer in
+        let parse_symbol_and_offset str =
+          try Scanf.sscanf str "%s@+0x%x" (fun symbol offset -> symbol, offset) with
+          | Scanf.Scan_failure _ | End_of_file -> "[unknown]", 0
+        in
+        let src_symbol, src_symbol_offset =
+          parse_symbol_and_offset src_symbol_and_offset
+        in
+        let dst_symbol, dst_symbol_offset =
+          parse_symbol_and_offset dst_symbol_and_offset
+        in
+        { thread =
+            { pid = (if pid = 0 then None else Some (Pid.of_int pid))
+            ; tid = (if tid = 0 then None else Some tid)
             }
-          | results ->
-            raise_s
-              [%message
-                "Regex of expected perf output did not match." (results : string array)])
+        ; time = time_lo + (time_hi * 1_000_000_000) |> Time_ns.Span.of_int_ns
+        ; kind =
+            (match String.strip kind with
+            | "call" -> Call
+            | "return" -> Return
+            | "tr strt" -> Start
+            | "tr end" -> End None
+            | "tr strt tr end" ->
+              (* CR-someday wduff: I saw this in practice, but I'm not sure what we want to do with
```

Comment:
Could you also file a ticket with some `perf script` output around this region, so we don't forget? I weakly think this should be treated as a jump, but we should check the perf code for what makes it generate this sequence.

Reply:
Could we collapse this as `| "tr strt  tr end" | kind` though, just to have only one error branch?

Reply:
Sure. I left them separate so to distinguish "we don't know what to do" from "we don't recognize this at all", but that seems kinda silly given that we're not being all that careful about error messages in the first place, and either way the fix is to go in and add support for the missing case.

I changed it to your suggestion.

Reply:
And opened issue #80 as well.